src/Ouroboros/Storage/Immutable/DB.hs

```diff
@@ -0,0 +1,418 @@
+    EpochFileNotFoundError cs -> "EpochFileNotFoundError: " <> prettyCallStack cs
+    IndexFileNotFoundError cs -> "IndexFileNotFoundError: " <> prettyCallStack cs
+    FileSystemError fs   -> prettyFSError fs
+    EpochIsReadOnlyError e1 e2 cs ->
+           "EpochIsReadOnlyError (input epoch was "
+        <> show e1
+        <> ", most recent epoch found: " <> show e2 <>"): "
+        <> prettyCallStack cs
+    InconsistentSlotError (RelativeSlot is) (RelativeSlot es) cs   ->
+           "InconsistentSlotError (input epoch was "
+        <> show is
+        <> ", expected was "
+        <> show es
+        <> "): "
+        <> prettyCallStack cs
+    SlotDoesNotExistError current requested cs   ->
+           "SlotDoesNotExistError (current is "
+        <> show current
+        <> ", requested was "
+        <> show requested
+        <> "): "
+        <> prettyCallStack cs
+
+renderFile :: String -> Epoch -> FsPath
+renderFile fileType epoch = [printf "%s-%03d.dat" fileType epoch]
+
+withDB :: (HasCallStack, MonadSTM m, MonadMask m, HasFSE m)
+       => FsPath
+       -> Epoch
+       -> (ImmutableDB m -> m (Either ImmutableDBError a))
+       -> m (Either ImmutableDBError a)
+withDB fsPath epoch action = runExceptT $
+    bracket (openDB fsPath epoch) closeDB (ExceptT . action)
+
+-- | Opens the database, creating it from scratch if the 'FilePath' points to
+-- a non-existing directory. Internally it preloads the most recent 'Epoch'.
+openDB :: (HasCallStack, MonadCatch m, MonadSTM m, HasFSE m)
+       => FsPath
+       -> Epoch
+       -> ExceptT ImmutableDBError m (ImmutableDB m)
+openDB fp currentEpoch = do
+      allFiles <- liftFsError $ do
+          createDirectoryIfMissing True fp
+          listDirectory fp
+      checkEpochConsistency allFiles
+      st    <- mkInternalState fp currentEpoch
+      stVar <- atomically $ newTMVar st
+      return $ ImmutableDB stVar fp
+  where
+      checkEpochConsistency :: (HasCallStack, MonadError ImmutableDBError m)
+                            => [String]
+                            -> m ()
+      checkEpochConsistency [] = return ()
+      checkEpochConsistency (x:xs) =
+          case parseFile x of
+               Nothing -> checkEpochConsistency xs
+               Just n  -> if n > currentEpoch
+                             then throwError (EpochIsReadOnlyError currentEpoch n callStack)
+                             else checkEpochConsistency xs
+
+      parseFile :: String -> Maybe Epoch
+      parseFile = readMaybe
+                . T.unpack
+                . snd
+                . T.breakOnEnd "-"
+                . fst
+                . T.breakOn "."
+                . T.pack
+
+
+-- | Closes the database, also closing any handle which have been opened in
+-- the process.
+closeDB :: (MonadSTM m, HasFSE m)
+        => ImmutableDB m -> ExceptT ImmutableDBError m ()
+closeDB ImmutableDB{..} = do
+    InternalState{..} <- atomically (takeTMVar _dbInternalState)
+    liftFsError $ do
```

Comment:
We can't ignore the error here but we should make sure to try and close both handles.